scenario/state.py

```diff
@@ -1760,8 +1774,86 @@ def _is_builtin_event(self, charm_spec: "_CharmSpec"):
         # assuming it is owned by the charm, LOOKS LIKE that of a builtin event or not.
         return self._path.type is not _EventType.custom
 
-    def deferred(self, handler: Callable, event_id: int = 1) -> DeferredEvent:
+    def _find_handler(self, charm_type: Type[CharmBase]) -> Callable:
```

Comment:
> I'm wondering whether we need to do ast parsing at all. Can't we inspect `Framework._observers` or something like that to know what handlers are attached to what event paths?

We could - we'd need to defer (sorry not sorry) figuring out what the handler was until we'd instantiated the Charm - until it was actually going to be saved, probably. That does tie in to your other suggestion (avoiding linking to the context in `_Event`), though.

I'll play with this a bit and see what I can come up with.

> (if I just destroyed your hour of fun and glory playing around with the ast, I'm sorry. I know how rare those occasions are...)

:smile: no problem - the fun is in the journey, right? :laughing: I'd already done much of it anyway, back when I was doing some static analysis of charms - looking at which events are most observed in this case.

> on a side note, that will also enable you to guess handlers from child Objects...!

It would indeed be nicer to have a solution that covers all the cases. That is possible with the ast method, just quite a bit more work to get right.